cpp-ch/local-engine/Storages/SubstraitSource/ReadBufferBuilder.cpp

```diff
@@ -118,87 +119,100 @@ class HDFSFileReadBufferBuilder : public ReadBufferBuilder
-            {
-                return start_pos;
-            }
-            size_t pos = start_pos;
-            int seek_status = hdfsSeek(fs, fin, pos);
-            if (seek_status != 0)
+            if (initial_pos == 0 || initial_pos == file_size)
+                return initial_pos;
+
+            int seek_ret = hdfsSeek(fs, fin, initial_pos);
+            if (seek_ret < 0)
+                throw DB::Exception(DB::ErrorCodes::CANNOT_SEEK_THROUGH_FILE, "Fail to seek HDFS file: {}, error: {}", file_path, std::string(hdfsGetLastError()));
+
+            static constexpr size_t buf_size = 1024;
+            char buf[buf_size];
+
+            auto do_read = [&]() -> int
             {
-                hdfsCloseFile(fs, fin);
-                throw DB::Exception(
-                    DB::ErrorCodes::CANNOT_SEEK_THROUGH_FILE,
-                    "Fail to seek HDFS file: {}, error: {}",
-                    file_path,
-                    std::string(hdfsGetLastError()));
-            }
-            char s[row_delimiter_size];
-            bool read_flag = true;
-            while (read_flag)
+                auto n = hdfsRead(fs, fin, buf, buf_size);
+                if (n < 0)
+                    throw DB::Exception(
+                        DB::ErrorCodes::CANNOT_READ_FROM_FILE_DESCRIPTOR,
+                        "Fail to read HDFS file: {}, error: {}",
+                        file_path,
+                        std::string(hdfsGetLastError()));
+
+                return n;
+            };
+
+            auto pos = initial_pos;
+            while (1)
             {
-                size_t read_size = hdfsRead(fs, fin, s, row_delimiter_size);
-                size_t i = 0;
-                for (; i < read_size && i < row_delimiter_size; ++i)
+                auto n = do_read();
+
+                /// If read to the end of file, return directly.
+                if (n == 0)
+                    return pos;
+
+                /// Search for \n or \r\n or \n\r in buffer.
+                int i = 0;
+                while (i < n)
                 {
-                    if (s[i] != *(row_delimiter.data() + i))
+                    if (buf[i] == '\n')
                     {
-                        break;
+                        if (i + 1 < n)
+                            return buf[i + 1] == '\r' ? pos + i + 2 : pos + i + 1;
+
+                        /// read again if buffer is not enough.
+                        auto m = do_read();
+                        if (m == 0)
+                            return pos + i + 1;
+
+                        return buf[0] == '\r' ? pos + i + 2 : pos + i + 1;
                     }
-                }
-                if (i == row_delimiter_size)
-                {
-                    char r[1];
-                    // The end of row maybe '\n', '\r\n', or '\n\r'
-                    if (hdfsRead(fs, fin, r, 1) != 0 && r[0] == '\r')
+                    else if (buf[i] == '\r')
                     {
```

Comment:
\r\n and \n can be handled by just judge whether the last character is \n

Reply:
This logic is from skipEndOfLine in src/Processors/Formats/Impl/CSVRowInputFormat.cpp